Allow user access to physical locations of mortar qpoints

lindsayad opened the pull request:
Closes #16177. Also refs #16171

@srinath-chakravarthy can you give this a try? Also I'm curious about your use case. If I  know what you're using it for, it might help me write a test for this (or maybe you want to add the test :smile:)

lindsayad committed:
Allow user access to physical locations of mortar qpoints

Closes #16177. Also refs #16171

srinath-chakravarthy commented:
@lindsayad thanks a lot. Will give this a try later tonight.
As for my use case, i was attempting to implement a spatially varying
GapConductance.
Let me see if i can come up with a test.

Cheers
Srinath

On Fri, Nov 13, 2020 at 2:55 PM Alex Lindsay <notifications@github.com>
wrote:

> Closes #16177 <https://github.com/idaholab/moose/issues/16177>. Also refs
> #16171 <https://github.com/idaholab/moose/discussions/16171>
>
> @srinath-chakravarthy <https://github.com/srinath-chakravarthy> can you
> give this a try? Also I'm curious about your use case. If I know what
> you're using it for, it might help me write a test for this (or maybe you
> want to add the test 😄)
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/idaholab/moose/pull/16182
> Commit Summary
>
>    - Allow user access to physical locations of mortar qpoints
>
> File Changes
>
>    - *M* framework/include/base/Assembly.h
>    <https://github.com/idaholab/moose/pull/16182/files#diff-0f1f2a521dd930c74ec62fcddfcc01c86f6dc7e38cd9845b328245544363e9f5>
>    (6)
>    - *M* framework/include/constraints/MortarConstraintBase.h
>    <https://github.com/idaholab/moose/pull/16182/files#diff-7345a9b3d222670aa12c4f58f967283bd34131dad796dbea8237031038ce84b4>
>    (3)
>    - *M* framework/src/constraints/MortarConstraintBase.C
>    <https://github.com/idaholab/moose/pull/16182/files#diff-79ee85e347e8d7387c6868af043d8847f24e45438cbb1ac691124a5c0553efa8>
>    (1)
>
> Patch Links:
>
>    - https://github.com/idaholab/moose/pull/16182.patch
>    - https://github.com/idaholab/moose/pull/16182.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/16182>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYC6LKSXZYKP6RTZNW7KI3SPWFJ3ANCNFSM4TU7TIDA>
> .
>

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/623347/) on 9705672 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16182/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/623354/) on 9705672 : invalidated by @lindsayad

lindsayad commented:
I'm just going to add some spatially varying source in the mortar constraint to test this

lindsayad committed:
Don't set interior parent or parent for mortar segment elements

It's unnecessary and actually just plain wrong. The elements we were
setting as interior parents and parents aren't even on the same mesh as
the mortar segment elements, so when we go to prepare the mortar segment
mesh for use, we get rightful complaints out of libmesh that we are
unable to determine which child of the parent element the mortar segment
elements are

lindsayad committed:
Make it so that MortarConstraints can be used without the LM var

This was always the intent, but apparently I never tested this
configuration before. With the new qoint test that I'm adding I really
don't have an LM variable

lindsayad committed:
Test _q_point user access

lindsayad commented:
Test failure is unrelated. Hopefully will be fixed by https://hpcgitlab.hpc.inl.gov/idaholab/pronghorn/-/merge_requests/638

moosebuild commented:
Job [App tests (Non Conda)](https://civet.inl.gov/job/625360/) on 9035a17 : invalidated by @lindsayad

Pronghorn test should be a little less brittle now

fdkong reviewed (approved):
Looks nice

fdkong commented on test/tests/mortar/mortar-q-points/test.i:
```diff
@@ -0,0 +1,69 @@
+
+[Problem]
+  kernel_coverage_check = false
+[]
```
Why you turned this off? Because block 20 and 10 do not have kernels?

lindsayad replied:
Hmm I may be able to remove this

lindsayad replied:
I actually cannot. In most mortar applications a user will have Lagrange Multiplier variable that lives on the secondary block, so I do have a coverage check that makes sure that there is variable defined on the secondary block (block '10' in this case; we do not check to see that there is a variable defined on the primary block). So if I remove this, then I get this error:
```
*** ERROR ***
Each subdomain must contain at least one Kernel.
The following block(s) lack an active kernel: 10 
```
However, it is a valid use case for mortar to not have an LM variable, but I would like users to have to explicitly say something in order for MOOSE not to error, in order to indicate that they know what they're doing.

fdkong replied:
I feel like the same thing can be implemented by the interface kernel as well. Using mortar is of course still good

lindsayad replied:
The domains are disconnected here (and even if they were the mesh would be non-conforming) so interface kernel would not work. Only beautiful mortar works here

fdkong commented on test/src/constraints/SpatiallyVaryingSource.C:
```diff
@@ -0,0 +1,42 @@
+  switch (type)
+  {
+    case Moose::MortarType::Secondary:
+      return _q_point[_qp] * tangent * _test_secondary[_i][_qp];
```
Just out of curiosity, I would expect you have something like f(`_q_point[_qp]`) instead of `_q_point[_qp].` The force depends on your coordination system. In other words, physics behaviors change if you choose a different coordination system.

lindsayad replied:
Yea, this is just a test object for making sure the values in `_q_point[_qp]` are correct. This definitely isn't realistic
